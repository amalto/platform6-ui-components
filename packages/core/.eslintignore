dist/
loader/

src/components.d.ts