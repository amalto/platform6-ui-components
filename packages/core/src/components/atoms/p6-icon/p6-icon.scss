$icon-dimensions: $size-normal !default;
$icon-dimensions-small: $size-small !default;

@import '@fortawesome/fontawesome-svg-core/styles';

:host {
  display: inline-flex;
  .icon {
    align-items: center;
    display: inline-flex;
    justify-content: center;
    height: $icon-dimensions;

    &.is-small {
      height: $icon-dimensions-small;
    }

    svg {
      height: 100%;
      width: 100%;
    }
  }
}
