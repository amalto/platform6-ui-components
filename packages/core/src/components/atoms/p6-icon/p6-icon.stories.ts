import { IconName, IconPrefix, library } from '@fortawesome/fontawesome-svg-core';
import { faAccessibleIcon, faAutoprefixer, faFontAwesome, faGit, faHtml5 } from '@fortawesome/free-brands-svg-icons';
import { faCat, faCheckSquare, faHome, faInfoCircle, faTimes } from '@fortawesome/free-solid-svg-icons';
import { Components } from '../../../components';
import { ComponentProps, getElement, getSelectArgType, makeSizeStory, makeStory, Props, StringSelectArgType } from '../../../shared/storybook';
import { Size } from '../../../shared/types';

const component = 'p6-icon';

export default {
  title: 'Atoms/Icon',
  component,
};

const componentProps: ComponentProps = ['name', 'size', 'iconPrefix', 'transform'];

const iconsMap = new Map([
  ['fab', ['accessible-icon', 'autoprefixer', 'font-awesome', 'git', 'html5']],
  ['fas', ['home', 'info-circle', 'check-square', 'times', 'cat']],
]);

library.add(
  // fab
  faAccessibleIcon,
  faAutoprefixer,
  faFontAwesome,
  faGit,
  faHtml5,
  // Fas
  faInfoCircle,
  faCheckSquare,
  faHome,
  faTimes,
  faCat,
);

const getStoryField = (name: IconName, props?: Props<Components.P6Icon>): HTMLElement => getElement(component, name, { name, ...props });

const makeLibStory = (iconPrefix: IconPrefix, defaultIcon: IconName) => {
  return makeStory<{
    icon: IconName;
    size: Size;
    transform: string;
  }>({
    componentProps,
    args: {
      icon: defaultIcon,
      size: Size.normal,
      transform: '',
    },
    argTypes: {
      ...getSelectArgType<StringSelectArgType>(
        'icon',
        (iconsMap.get(iconPrefix) || []).map(type => ({
          key: type,
          value: type,
        })),
      ),
    },
    builder: ({ icon, ...args }): HTMLElement => getStoryField(icon, { iconPrefix, ...args }),
  });
};

export const LibraryFas = makeLibStory('fas', 'home');
export const LibraryFab = makeLibStory('fab', 'cat');

export const Sizes = makeSizeStory({
  componentProps,
  builder: ({ value }) =>
    getStoryField('home', {
      size: value,
    }),
});

export const HomeIcon = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField('home'),
});

export const FlipAndRotate = makeStory<{ transform: string }>({
  componentProps,
  args: {
    transform: 'flip-v rotate-90',
  },
  builder: (props): HTMLElement => getStoryField('home', props),
});
