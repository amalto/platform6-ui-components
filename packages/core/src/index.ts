export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
