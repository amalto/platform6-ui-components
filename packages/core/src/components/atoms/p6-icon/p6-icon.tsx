import { icon, IconName, IconPrefix, parse } from '@fortawesome/fontawesome-svg-core';
import { Component, Element, h, JSX, Prop } from '@stencil/core';
import { Size } from '../../../shared/types';
import { cleanupAttributes, isEmpty } from '../../../utils/attribute';
import { getSizeClass } from '../../../utils/classes';

export type IconStyle = IconPrefix;

@Component({
  tag: 'p6-icon',
  styleUrl: 'p6-icon.scss',
  shadow: true,
})
export class P6Icon {
  @Element() host!: HTMLP6IconElement;

  /**
   * Icon name
   */
  @Prop() name!: IconName;

  /**
   * set the size of the button
   */
  @Prop() size: Size = Size.normal;

  /**
   * Style prefix
   */
  @Prop({ attribute: 'iconPrefix' }) iconPrefix: IconPrefix = 'fas';

  /**
   * transformation performed on the icon.
   */
  @Prop() transform: string | undefined;

  render(): JSX.Element | null {
    const transform = isEmpty(this.transform) ? undefined : parse.transform(this.transform as string);

    const img = icon({ prefix: this.iconPrefix, iconName: this.name }, cleanupAttributes({ transform }));

    if (img === undefined) {
      // eslint-disable-next-line no-console
      console.debug(`p6-icon : could not find icon ${this.name} (style: ${this.iconPrefix})`);
      return null;
    }

    const classes = {
      icon: true,
      ...getSizeClass(this.size),
    };

    return <span class={classes} innerHTML={img.html[0]} />;
  }
}
