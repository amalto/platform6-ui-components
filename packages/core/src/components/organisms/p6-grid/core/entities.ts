import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
