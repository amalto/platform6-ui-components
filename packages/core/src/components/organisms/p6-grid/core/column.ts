import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
