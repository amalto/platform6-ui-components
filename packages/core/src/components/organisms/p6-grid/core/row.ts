import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
