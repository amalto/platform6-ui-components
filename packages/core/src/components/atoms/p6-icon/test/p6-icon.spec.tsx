import { IconDefinition, library } from '@fortawesome/fontawesome-svg-core';
import { newSpecPage } from '@stencil/core/testing';
import { P6Icon } from '../p6-icon';

const getIcon: () => IconDefinition = () => ({
  prefix: 'fas',
  iconName: 'chart-bar',
  icon: [
    512,
    512,
    [],
    'f080',
    'M332.8 320h38.4c6.4 0 12.8-6.4 12.8-12.8V172.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V76.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-288 0h38.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zM496 384H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-32c0-8.84-7.16-16-16-16z',
  ],
});

library.add(getIcon());

describe('p6-icon', () => {
  it('renders an icon', async () => {
    const icon = getIcon();

    const page = await newSpecPage({
      components: [P6Icon],
      html: `<p6-icon name="chart-bar"></p6-icon>`,
    });

    const svg = page.root?.shadowRoot?.querySelector('svg');
    expect(svg).not.toBeNull();
    expect(svg?.getAttribute('data-icon')).toEqual(icon.iconName);
    expect(svg?.getAttribute('data-prefix')).toEqual(icon.prefix);

    const path = svg?.querySelector('path');
    expect(path).toBeDefined();
    expect(path?.getAttribute('d')).toEqual(icon.icon.reverse().shift());
  });

  it('renders nothing and add a debug message when the icon is unknown', async () => {
    const spyDebug = jest.fn();
    Object.defineProperty(console, 'debug', {
      writable: true,
      value: spyDebug,
    });

    const page = await newSpecPage({
      components: [P6Icon],
      html: `<p6-icon name="unknown-icon"></p6-icon>`,
    });

    const svg = page.root?.shadowRoot?.querySelector('svg');
    expect(svg).toBeNull();
    expect(spyDebug).toHaveBeenCalledTimes(1);

    const message = spyDebug.mock.calls[spyDebug.mock.calls.length - 1][0];
    expect(message).toContain('p6-icon');
    expect(message).toContain('unknown-icon');
    expect(message).toContain('fas');
  });

  it('can transform an icon', async () => {
    const page = await newSpecPage({
      components: [P6Icon],
      html: `<p6-icon name="chart-bar" transform="flip-v rotate-90"></p6-icon>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
