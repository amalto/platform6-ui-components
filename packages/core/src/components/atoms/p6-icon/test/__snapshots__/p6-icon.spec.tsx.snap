// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-icon can transform an icon 1`] = `
<p6-icon name="chart-bar" transform="flip-v rotate-90">
  <mock:shadow-root>
    <span class="icon">
      <svg aria-hidden="true" class="fa-chart-bar fa-w-16 svg-inline--fa" data-icon="chart-bar" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg" style="transform-origin: 0.5em 0.5em;">
        <g transform="translate(256 256)">
          <g transform="translate(0, 0)  scale(1, -1)  rotate(90 0 0)">
            <path d="M332.8 320h38.4c6.4 0 12.8-6.4 12.8-12.8V172.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V76.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-288 0h38.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zM496 384H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-32c0-8.84-7.16-16-16-16z" fill="currentColor" transform="translate(-256 -256)"></path>
          </g>
        </g>
      </svg>
    </span>
  </mock:shadow-root>
</p6-icon>
`;
